classes/submission/SubmissionDAO.inc.php

```diff
@@ -94,30 +95,110 @@ function newDataObject() {
 		$submission->setLastModified($this->datetimeFromDB($row['last_modified']));
 		$submission->setLanguage($row['language']);
 		$submission->setCitations($row['citations']);
 
-		$this->getDataObjectSettings('submission_settings', 'submission_id', $submission->getId(), $submission);
+		$this->getDataObjectSettings('submission_settings', 'submission_id', $submission->getId(), $submission, $submissionRevision, 'submission_revision');
 
 		return $submission;
 	}
 
 	/**
+	 * Get all revisions for a submission
+	 * @param $submissionId int
+	 * @param $contextId int optional
+	 * $param $order string optional default: ASC
+	 * @return array
+	 */
+	function getSubmissionRevisionIds($submissionId, $contextId = null, $order = SORT_DIRECTION_ASC) {
+
+		if (!ctype_digit("$submissionId")) {
+			$publishedArticleDao = DAORegistry::getDAO('PublishedArticleDAO');
+			$publishedArticle = $publishedArticleDao->getPublishedArticleByPubId('publisher-id', $submissionId, $contextId);
+			$submissionId = $publishedArticle->getId();
+		}
+
+		$params = array((int) $submissionId);
+		if ($contextId) {
+			$params[] = (int) $contextId;
+		}
+
+		$sql = 'SELECT DISTINCT submission_revision FROM submission_settings ' .
+			   'INNER JOIN submissions ON submission_settings.submission_id = submissions.submission_id' .
+			   ' WHERE submission_settings.submission_id = ?'  .
+			   ($contextId ? ' AND submissions.context_id = ?' : '') .
+			   ' ORDER BY submission_revision ' . $this->getDirectionMapping($order);
+
+		$result = $this->retrieve($sql, $params);
+		$submissionRevisions = array();
+
+		foreach($result->getArray() as $submission) {
+			$submissionRevisions[] = $submission['submission_revision'];
+		}
+
+		return $submissionRevisions;
+	}
+
+	/**
+	 * Get all published revisions for a submission
+	 * @param $submissionId int
+	 * @return array
+	*/
+	function getPublishedSubmissionRevisions($submissionId, $contextId = null, $order = SORT_DIRECTION_DESC){
+		$submissionRevisions = $this->getSubmissionRevisionIds($submissionId, $contextId, $order);
+		$publishedSubmissionRevisions = array();
+
+		foreach($submissionRevisions as $revision){
+			$submission = $this->getById($submissionId, null, false, $revision);
+			if($submission->getDatePublished()){
+				$publishedSubmissionRevisions[] = $submission;
+			}
+		}
+
+		return $publishedSubmissionRevisions;
+
+	}
+
+	/**
+	 * Get the current revision id for a submission
+	 * @param $submissionId int
+	 * @param $contextId int
+	 * @return int
+	 */
+	function getLatestRevisionId($submissionId, $contextId = null) {
+		$submissionRevisions = $this->getSubmissionRevisionIds($submissionId, $contextId, SORT_DIRECTION_ASC);
+		return array_pop($submissionRevisions);
+	}
+
+	/**
+	 * Get the current published revision id for a submission
+	 * @param $submissionId int
+	 * @param $contextId int
+	 * @return int
+	 */
+	function getLatestPublishedRevisionId($submissionId, $contextId = null) {
```

Comment:
This function seem to be used only one time, in pages/article/ArticleHandler.inc.php, where the function getPublishedSubmissionRevisions is also called before this function... thus I think we do not need this function for now and can do array_pop in ArticleHandler